src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -87,115 +109,119 @@ EmbeddedAssemblies::zip_load_entries (int fd, const char *apk_name, monodroid_sh
 				continue;
 			}
 		}
 #endif // def NET6
 
 		// assemblies must be 4-byte aligned, or Bad Things happen
 		if ((data_offset & 0x3) != 0) {
-			log_fatal (LOG_ASSEMBLY, "Assembly '%s' is located at bad offset %lu within the .apk\n", file_name, data_offset);
+			log_fatal (LOG_ASSEMBLY, "Assembly '%s' is located at bad offset %lu within the .apk\n", entry_name.get (), data_offset);
 			log_fatal (LOG_ASSEMBLY, "You MUST run `zipalign` on %s\n", strrchr (apk_name, '/') + 1);
 			exit (FATAL_EXIT_MISSING_ZIPALIGN);
 		}
 
-		bool entry_is_overridden = !should_register (strrchr (file_name, '/') + 1);
+		const char *last_slash = utils.find_last (entry_name, '/');
+		bool entry_is_overridden = last_slash == nullptr ? false : !should_register (last_slash + 1);
 
-		if ((utils.ends_with (file_name, ".pdb") || utils.ends_with (file_name, ".mdb")) &&
+		if ((utils.ends_with (entry_name, ".pdb") || utils.ends_with (entry_name, ".mdb")) &&
 				register_debug_symbols &&
 				!entry_is_overridden &&
-				bundled_assemblies != nullptr) {
-			md_mmap_info map_info = md_mmap_apk_file(fd, data_offset, file_size, file_name, apk_name);
-			if (register_debug_symbols_for_assembly (file_name, (bundled_assemblies) [bundled_assemblies_count - 1], (const mono_byte*)map_info.area, static_cast<int>(file_size)))
+				bundled_assembly_index >= 1) {
+			md_mmap_info map_info = md_mmap_apk_file(fd, data_offset, file_size, entry_name.get (), apk_name);
+			if (register_debug_symbols_for_assembly (entry_name, (bundled_assemblies) [bundled_assembly_index - 1], (const mono_byte*)map_info.area, static_cast<int>(file_size)))
 				continue;
 		}
 
 #if !defined(NET6)
-		if (utils.ends_with (file_name, ".config") && bundled_assemblies != nullptr) {
-			char *assembly_name = strdup (basename (file_name));
+		if (utils.ends_with (entry_name, ".config") && !bundled_assemblies.empty ()) {
+			char *assembly_name = strdup (basename (entry_name.get ()));
 			// Remove '.config' suffix
 			*strrchr (assembly_name, '.') = '\0';
 
-			md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, file_name, apk_name);
+			md_mmap_info map_info = md_mmap_apk_file (fd, data_offset, file_size, entry_name.get (), apk_name);
 			mono_register_config_for_assembly (assembly_name, (const char*)map_info.area);
 
 			continue;
 		}
 #endif // ndef NET6
 
-		if (!utils.ends_with (file_name, ".dll"))
+		if (!utils.ends_with (entry_name, ".dll"))
 			continue;
 
 		if (entry_is_overridden)
 			continue;
 
-		size_t alloc_size = MULTIPLY_WITH_OVERFLOW_CHECK (size_t, sizeof(void*), bundled_assemblies_count + 1);
-		bundled_assemblies = reinterpret_cast<MonoBundledAssembly**> (utils.xrealloc (bundled_assemblies, alloc_size));
-		MonoBundledAssembly *cur = bundled_assemblies [bundled_assemblies_count] = reinterpret_cast<MonoBundledAssembly*> (utils.xcalloc (1, sizeof (MonoBundledAssembly)));
-		++bundled_assemblies_count;
+		assembly_count--;
```

Comment:
I look at this and wonder…. This statement is "weird" even in-context; why is it being decremented?  It starts as `application_config.number_of_assemblies_in_apk`, which could be 0 for a "debug+fastdev" build…

but why decrement *first*?  Why not instead change the `if` below to:

```cpp
if (XA_UNLIKELY (assembly_count == 0 || …)
```

I don't understand why this decrement is needed in the first place.  Why this particular pattern?

Reply:
It's a "low level" marker. And the reason for decrementing is that it's the cheapest approach here. Your proposal of `assembly_count == 0` check means that we'd have to stop decrementing at 0, which means changing line 183 to

```cpp
if (assembly_count > 0)
  assembly_count--;
```

which means a tiny, and unnecessary, performance loss. And it is first decremented because vector indices are 0-based and the counter starts from the vector **size** (`application_config.number_of_assemblies_in_apk`) and decrementing it keeps it in "sync" with `bundled_assembly_index` which counts up from a (valid) index of 0, thus the decrement maintains the symmetry. While `assembly_count` isn't used for indexing, I need it to hit the low level marker (-1) at the same time `bundled_assembly_index` becomes larger than the vector size.